crates/wasmtime/src/runtime/vm/cow.rs

```diff
@@ -528,13 +540,33 @@ impl MemoryImageSlot {
     #[allow(dead_code)] // ignore warnings as this is only used in some cfgs
     unsafe fn reset_with_original_mapping(
         &mut self,
-        keep_resident: usize,
+        keep_resident: HostAlignedByteCount,
         mut decommit: impl FnMut(*mut u8, usize),
     ) {
         match &self.image {
             Some(image) => {
-                assert!(self.accessible >= image.linear_memory_offset + image.len);
-                if image.linear_memory_offset < keep_resident {
+                let image_end = image
+                    .linear_memory_offset
+                    .checked_add(image.len)
+                    .expect("image is in bounds");
+                let mem_after_image = self
+                    .accessible
+                    .checked_sub(image_end)
+                    .expect("image_end falls before self.accessible");
+
+                let keep_resident_minus_image_start =
+                    if let Ok(excess) = keep_resident.checked_sub(image.linear_memory_offset) {
+                        // The .is_zero() check is carried over from a previous
+                        // iteration of the logic, which had > 0 and <= 0 cases.
+                        //
+                        // XXX: are the two paths identical? If so, the is_zero
+                        // check can go away.
```

Comment:
Thoughts about this? I tried to be super cautious about signs this time.

Reply:
Would it be possible to keep the same `<` condition as before? I find that one a bit easier to understand in terms of the diagrams below personally.

Reply:
Done.